[Go] Fix performance issues with expired cache (decrypt) - issue #385

keperry opened the pull request:
This should fix issue #385 
It appears that when the cache item was expired and then reloaded, it was never written back to the cache.  This caused the item to get loaded every time after RevokeCheckInterval  expired.

There is a small fix for writing back to the cache, and then I also moved the locks around to load/get.   This should allow for better performance on reads with multiple readers.  I also ran "go mod tidy".  I added a benchmark test and normal test so we can test this in the future. 
To run the benchmark, try something like this:
`go test -run="notthis" -bench='^BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey$' -v -cpu=12 -cpuprofile=cpu.out -benchtime=60s 
`
I noticed significant improvements after making the "write" fix.  After making the "write" fix + "lock changes", I found a slight improvement as well.

CPU profiles:

- Original: `BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey-12           12675           5642228 ns/op`

- With Write Fix Only: `BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey-12        98069263              1094 ns/op`

- With Write Fix And Lock Changes: `BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey-12        100000000              803.6 ns/op`

I have attached the [cpuprofiles.zip](https://github.com/godaddy/asherah/files/6849304/cpuprofiles.zip)
 for all three scenarios.

1. Original
2. Write fix only
3. Write fix with lock changes

Here we are comparing the original with the write fix only:
`go tool pprof -nodefraction=0 -http=:8086 -normalize -diff_base cpu_original.out cpu_write_fix.out 
`![cpu_diff_write_fix](https://user-images.githubusercontent.com/1618767/126345466-e93dfbd0-ceec-40fb-952a-23bf6bc7a275.png)


Here is comparing write fix with write fix and lock changes : 
`go tool pprof -nodefraction=0 -http=:8086 -normalize -diff_base cpu_write_fix.out cpu_write_fix_with_lock_changes.out 
`![image](https://user-images.githubusercontent.com/1618767/126343817-5a78aa73-d9fd-413c-81fa-b83d19a5fee6.png)

We ran a small canary test in prod.  With the RevokeCheckInterval set to 5 minutes, we saw much better performance with this fix.
![Datafabric_DRR_metrics_after_fix](https://user-images.githubusercontent.com/1618767/126345724-95f1935a-77af-4136-baa8-84353d5b1dec.png)

aka-bo commented on go/appencryption/cache.go:
```diff
@@ -96,9 +96,6 @@ func isReloadRequired(entry cacheEntry, checkInterval time.Duration) bool {
 // is not present in the cache it will retrieve the key using the provided keyLoader
 // and store the key if an error is not returned.
 func (c *keyCache) GetOrLoad(id KeyMeta, loader keyLoader) (*internal.CryptoKey, error) {
-	c.rw.Lock()
-	defer c.rw.Unlock()
-
```
While locking at this level does incur a performance cost (as you've found) it is necessary. Moving the lock to the get/lock methods introduces a deadlock scenario.

aka-bo commented on go/appencryption/cache.go:
```diff
@@ -112,6 +109,9 @@ func (c *keyCache) GetOrLoad(id KeyMeta, loader keyLoader) (*internal.CryptoKey,
 // The second return value indicates the successful retrieval of a
 // fresh key.
 func (c *keyCache) get(id KeyMeta) (*internal.CryptoKey, bool) {
+	c.rw.RLock()
+	defer c.rw.RUnlock()
+
```
as noted above, this needs to be reverted.

aka-bo commented on go/appencryption/cache.go:
```diff
@@ -127,6 +127,9 @@ func (c *keyCache) get(id KeyMeta) (*internal.CryptoKey, bool) {
 // load maintains the latest entry for each distinct ID which can be accessed using
 // id.Created == 0.
 func (c *keyCache) load(id KeyMeta, loader keyLoader) (*internal.CryptoKey, error) {
+	c.rw.Lock()
+	defer c.rw.Unlock()
+
```
revert

aka-bo reviewed (changes requested):
Hi @keperry and thank you for the PR and excellent write-up describing the problem!

The underlying fix looks good but as noted in the comments the locking changes need to be reverted as they introduce a race (deadlock).

keperry commented:
Note:  Modified locks to use RLock and Lock on GetOrLoad().  This should get us both quick reads and performant writes.  I also added an early exit for the heavy "Lock()" - that should prevent several queued up RW locks from all performing loads.   Here are the profiles for both scenarios:  
[cpuprofiles_take2.zip](https://github.com/godaddy/asherah/files/6851021/cpuprofiles_take2.zip)


Here are some results from my machine - the performance looks a lot better with the lock changes:
- Write Fix only: 
`BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey-12        129446646              585.7 ns/op`
- Write Fix with Lock Improvements:
`BenchmarkKeyCache_GetOrLoad_MultipleThreadsRead_NeedReloadKey-12        551315848              177.4 ns/op
`
![image](https://user-images.githubusercontent.com/1618767/126390859-952386b5-054b-4038-bcf5-c083651d91d9.png)

aka-bo reviewed (approved):
LGTM

aka-bo commented:
Merging... I'll take a closer look at the unrelated build issue as part of the release process.

Thanks again for the PR!
